### PR TITLE
Building AWS Dashboard Console using Arduino -initial commit

### PR DESCRIPTION
**Pull request by m1kma**
Rakensin AWS dashboardin Arduinolla. Tässä postaus aiheesta.

### CODE REVIEW

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+layout: post
+title: Building AWS Dashboard Console using Arduino
+author: m1kma
+excerpt: Arduino compatible microcontrollers are an easy and fun way to create electronic projects. In this blogpost I will cover details of the Arduino project to create a AWS dashboard console prototype.  
+tags:
+- Arduino
+- Electronics
+- AWS
+---
+The system dashboard plays an important role when a real-time system status monitoring and rapid detection of error situations are required. It presents a general view of the running system including metrics, alarms and historical timeseries data. In addition, the dashboard could present special information such as a deployment pipeline status, which is useful especially for the software developers.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: In addition, the dashboard... kuulostaa oudolta mun korvaan. Mun mielestä pitäisi olla In addition, a dashboard...

Ja a deployment pipeline status -> artikkeli pois.

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+Traditional dashboards are usually displayed on computer or TV screen. Therefore, the dashboard must be displayed on the screen to monitor the system status. Is it really necessary? Is there any other way to monitor the system health? An alternative to solve this problem is to build a separate electronic device that is on the table. This device can display the system status f. ex. by lights. The device is an effective way to present system status events immediately without having to see the display. 
+
+![Dashboard console](/img/aws-dashboard-console-using-arduino/dashboard-console.jpg "Dashboard console")
+
+## The AWS Dashboard Console Project
+
+The aim of this project was to create a prototype of the AWS dashboard console using Arduino compatible board and electronic components. The project was mainly a hobby project, but the idea behind was also to create something that I could use on my daily work.
+
+The console presents AWS alerts and codepipeline status by the LED lights and LCD screen. The console runs independently by polling AWS metrics over the WiFi. The console uses NodeMCU microcontroller with an integrated ESP8266 WiFi chip. Just plug-in the power and go.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: Voisko laittaa jotkut linkit noihin NodeMCH ja ESP8266 WiFi chip kohtiin? :)

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+## The AWS Dashboard Console Project
+
+The aim of this project was to create a prototype of the AWS dashboard console using Arduino compatible board and electronic components. The project was mainly a hobby project, but the idea behind was also to create something that I could use on my daily work.
+
+The console presents AWS alerts and codepipeline status by the LED lights and LCD screen. The console runs independently by polling AWS metrics over the WiFi. The console uses NodeMCU microcontroller with an integrated ESP8266 WiFi chip. Just plug-in the power and go. 
+
+![Dashboard architecture](/img/aws-dashboard-console-using-arduino/dashboard-architecture.png "Dashboard architecture")
+
+The architecture contains a NodeMCU board and AWS Lambda backend. The lambda serves AWS status information via the REST interface to the internet. The console polls the REST interface periodically and controls LED lights and LCD screen.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: sanoisin, via a REST interface.

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+- 1 x NodeMCU board
+- 1 x LCD 2x16 characters I2C
+- 7 x LED lights
+- 7 x resistors for the LEDs
+- 1 x breadboard
+- wires
+
+![Dashboard circuit](/img/aws-dashboard-console-using-arduino/dashboard-circuit.png "Dashboard circuit")
+
+Basic layout of the circuit is a relative simple. The LEDs (and resistors) are connected to the NodeMCU IO pins. The LCD screen data wires are connected to the default I2C IO pins. The LCD screen power is connected to the NodeMCU 5V VIN pin. The NodeMCU board is powered by the USB connector. Components are assembled on the breadboard to minimize soldering.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: relatively simple

Ehkäpä: ..minimize soldering required.

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+> *Note that this circuit contains a conflict: the NodeMCU board runs by an 3V internal voltage and IO pins are therefore 3V tolerant. On opposite, the LCD screen requires a 5V voltage. Related to the documentation, this circuit should not work since the 5V LCD data wires are connected to the 3V NodeMCU pins. Luckily I found blog posts where this kind of the circuit was proven to work. At least my version of the NodeMCU seems to work fine event the 3V tolerant IO pins are driven by the 5V voltage.*
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: Oisko tohon kiva laittaa linkki tuohon blogipostaukseen?

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    
+    // Set the console
+    setConsole(data_dev, data_test, data_prod);        
+  } else {
+    lcd.print("Wifi Conn lost   ");
+  }
+}
+```
+
+The each AWS environment (DEV, TEST and PROD) contains own Lambda script to serve status information. Environments are call separately and results are passed to the `setConsole` method that controls the electronics.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: The each AWS... -> Each AWS...
contains it's own Lambda..
are called separately

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    // Set the console
+    setConsole(data_dev, data_test, data_prod);        
+  } else {
+    lcd.print("Wifi Conn lost   ");
+  }
+}
+```
+
+The each AWS environment (DEV, TEST and PROD) contains own Lambda script to serve status information. Environments are call separately and results are passed to the `setConsole` method that controls the electronics. 
+> *In this context the environment is equal with an AWS account.*
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: a environment

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    }
+  }
+
+  String line = client.readStringUntil('}');
+  line = line + "}";
+
+  client.stop();
+```
+
+The low level library `WiFiClientSecure` is used for the HTTP request. High level HTTP libraries are also available but I wanted to stay in basics and decided to use this low level implementation. The HTTP response is read until a closing parenthesis which is a last character of the HTTP REST response.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: until closing parenthesis is encountered tms.

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+Few examples of the LCD output:
+
+![Dashboard LCD](/img/aws-dashboard-console-using-arduino/dashboard-lcd.png "Dashboard LCD")
+
+If the string is longer than 16 characters (LCD width), then it is cut. 
+
+>*The LCD library has an extra option to scroll overflowing content on the screen but I note that it was not suitable for this case. Scrolling every string on the screen would take just too much time.*
+
+### Lambda backend script
+AWS status information is served by the simple Lambda script (Python). The Lambda provides a REST interface that includes all necessary information for the console. The Lambda starts on request and does not contain persistence. The Lambda is publised though the AWS API Gateway and secured by an API KEY.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: publised -> published
though -> through

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+## Summary
+
+The project contained few stages:
+1. Planning the required components
+2. Prototyping a first circuit version using breadboard
+3. Coding the program
+4. Building a final version to the box and solder the required wiring
+
+The most laborious stage was building the box and place all components to it. I found the plastic box from the local electronic shop and made the required holes by a drill and knife. That took a surprising amount of time and the result is still a bit rough.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: surprisingly large amount of time

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+The coding was relatively easy but there was still some troubles on the road. Originally I used an old version of the JSON parser that affected issues. The old version was using C++ pointers that I had never used before. The pointers affected a strange behaviour when the JSON properties was print to the LCD. After updating a latest version of the JSON parser, everything worked fine.
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: The coding kuulostaa oudolta. Laittaisin:
Programming was relatively easy, but there (huom, pilkku) myös along the road.

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+My original plan was to add fancy mechanical gauges and servo motors to the console. I gave up that idea because I was a bit lazy and didn’t figure out what kind of benefit the gauges and servos would provide. I still have an idea about the gauges so maybe I will create a second version in future with a steampunk style :)
```
Comment (juhofriman) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: in the future in a steampunk style

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+if (pipelines_running_test == true) { digitalWrite(PIPE_RUNNING_LED_TEST, HIGH); }
+if (pipelines_failed_test == true) { digitalWrite(PIPE_FAILED_LED_TEST, HIGH); }
+if (alarms_raised_prod == true) { digitalWrite(ALARMS_LED, HIGH); }
+if (pipelines_running_prod == true) { digitalWrite(PIPE_RUNNING_LED_PROD, HIGH); }
+if (pipelines_failed_prod == true) { digitalWrite(PIPE_FAILED_LED_PROD, HIGH); } 
+```
+
+First, the LEDs are switch Off to clear the previous state. Then the necessary LEDs are switch On by the simple `if` clauses.
+
+**After the LED’s are fine, write content to the LCD:**
```
Comment (mpajunen) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: LEDs

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+      lcd.print("     All OK     ");
+
+      delay(60000);
+
+      break;
+    }
+  }
+```
+
+Every `if` clause contains 5 second delay to keep the certain text on the screen for a moment. The `for` loop is used to “blink” all texts few times on the screen. If there are nothing to show, then print “All OK”.
```
Comment (mpajunen) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: If there **is** nothing

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 15]
```diff
@@ -0,0 +1,265 @@
+---
+layout: post
+title: Building AWS Dashboard Console using Arduino
+author: m1kma
+excerpt: Arduino compatible microcontrollers are an easy and fun way to create electronic projects. In this blogpost I will cover details of the Arduino project to create a AWS dashboard console prototype.  
+tags:
+- Arduino
+- Electronics
+- AWS
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: Pitäiskö tageihin lisätä seksiä lisää hakukonenäkyvyysmielessä: devops, monitoring, jne?

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+
+The console presents AWS alerts and codepipeline status by the LED lights and LCD screen. The console runs independently by polling AWS metrics over the WiFi. The console uses NodeMCU microcontroller with an integrated ESP8266 WiFi chip. Just plug-in the power and go. 
+
+![Dashboard architecture](/img/aws-dashboard-console-using-arduino/dashboard-architecture.png "Dashboard architecture")
+
+The architecture contains a NodeMCU board and AWS Lambda backend. The lambda serves AWS status information via the REST interface to the internet. The console polls the REST interface periodically and controls LED lights and LCD screen.
+
+### Arduino and NodeMCU/ESP8266 microcontroller
+
+[Arduino](https://www.arduino.cc) series boards, such as a popular Arduino Uno, contains a simple programmable microprocessor and set of GPIO connectors for the electronic components. Some of the board models has also an integrated WiFi chip which enables the internet connection. The Arduino compatible boards do not include an operating system which is the main difference compared to the popular Rasberry Pi series boards.
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: Rasberry --> Raspberry

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 1 above)
Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: timeseries --> time series/time-series ?

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+
+Traditional dashboards are usually displayed on computer or TV screen. Therefore, the dashboard must be displayed on the screen to monitor the system status. Is it really necessary? Is there any other way to monitor the system health? An alternative to solve this problem is to build a separate electronic device that is on the table. This device can display the system status f. ex. by lights. The device is an effective way to present system status events immediately without having to see the display. 
+
+![Dashboard console](/img/aws-dashboard-console-using-arduino/dashboard-console.jpg "Dashboard console")
+
+## The AWS Dashboard Console Project
+
+The aim of this project was to create a prototype of the AWS dashboard console using Arduino compatible board and electronic components. The project was mainly a hobby project, but the idea behind was also to create something that I could use on my daily work.
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: on my daily work --> in my daily work?

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 2 above)
Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: plug-in --> plug in

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+
+### Arduino and NodeMCU/ESP8266 microcontroller
+
+[Arduino](https://www.arduino.cc) series boards, such as a popular Arduino Uno, contains a simple programmable microprocessor and set of GPIO connectors for the electronic components. Some of the board models has also an integrated WiFi chip which enables the internet connection. The Arduino compatible boards do not include an operating system which is the main difference compared to the popular Rasberry Pi series boards.
+
+[NodeMCU](http://www.nodemcu.com/index_en.html) is an Arduino compatible open source IoT board. It includes hardware which is based on the ESP8266/ESP-12 WiFi module. The NodeMCU is comparable to the common Arduino boards; it can be programmed by the Arduino IDE and it *stupidly* runs the code that is stored to its flash memory.
+
+### Circuit of the Console
+
+The circuit of the console contains following components:
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: following --> the following

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 4 above)
Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: Basic layout of the circuit is a relative simple. --> The basic layout of the circuit is relatively simple.

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 5 above)
Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: On opposite -> On the opposite

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+  lcd.setCursor(0, 0);
+  lcd.print("Call AWS...     ");
+  lcd.setCursor(0, 1);
+  lcd.print("                ");
+}
+```
+
+First, the LED pins are defined. This definition is required to setup which IO pins the LED wires are connected. Then the IO pin mode is set to “OUTPUT” which is required to enable a write mode for the pins.
+
+The LCD screen is initialized and backlight is switch on. The electronic components are initialized by the custom method `initConsole`. Basically the `initConsole` will blink the LEDs and write some text to the screen. Finally, the WiFi is connected.
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: and backlight is switch on --> and the backlight is switched on?

Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: Basically the `initConsole` will blink -->
Basically, the `initConsole` will blink
Tai pelkästään 
The `initConsole` will blink

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 6 above)
Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: DEV, TEST and PROD --> DEV, TEST, and PROD

Environments are call --> Environments are called

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 8 above)
Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: The low level  --> The low-level

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 14 above)
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: 5 second delay --> 5-second delay 
If there are nothing to show --> If there is nothing to show

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 9 above)
Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: necessary --> the necessary

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+    "pipelines_running": false,
+    "pipelines_running_list": [],
+    "pipelines_failed": false,
+    "pipelines_failed_list": []
+}
+```
+
+## Summary
+
+The project contained few stages:
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: contained few --> contained a few

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+1. Planning the required components
+2. Prototyping a first circuit version using breadboard
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: using breadboard --> using a breadboard

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 11 above)
Comment (crystoll) [reply _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: when the JSON properties was print to --> when the JSON properties were printed to
After updating a latest version --> After updating the latest version

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+3. Coding the program
+4. Building a final version to the box and solder the required wiring
+
+The most laborious stage was building the box and place all components to it. I found the plastic box from the local electronic shop and made the required holes by a drill and knife. That took a surprising amount of time and the result is still a bit rough.
+
+The coding was relatively easy but there was still some troubles on the road. Originally I used an old version of the JSON parser that affected issues. The old version was using C++ pointers that I had never used before. The pointers affected a strange behaviour when the JSON properties was print to the LCD. After updating a latest version of the JSON parser, everything worked fine.
+
+My original plan was to add fancy mechanical gauges and servo motors to the console. I gave up that idea because I was a bit lazy and didn’t figure out what kind of benefit the gauges and servos would provide. I still have an idea about the gauges so maybe I will create a second version in future with a steampunk style :)
+
+Early prototype of the dashboard:
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: An Early..

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+
+![Dashboard early proto](/img/aws-dashboard-console-using-arduino/dashboard-early-proto.jpg "Dashboard early proto")
+
+On the following video couple AWS Pipelines are started:
+
+[![Running pipelines](http://img.youtube.com/vi/fGvGPpxiK4g/0.jpg)](http://www.youtube.com/watch?v=fGvGPpxiK4g)
+
+### Price of the components
+
+Total price of the components was about 40 euros. The most expensive parts where:
```
Comment (crystoll) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: The total price..
bredboard --> breadboard

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 10 above)
Comment (mpajunen) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: and plac**ing** all components
**with** a drill and **a** knife

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 11 above)
Comment (mpajunen) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: JSON parser that **caused** issues
The pointers **caused**
properties **were** print**ed**

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 12 above)
Comment (mpajunen) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: and **couldn't** figure out

File: _posts/2019-04-10-aws-dashboard-console-using-arduino.md (hunk 23 above)
Comment (mpajunen) [inline _posts/2019-04-10-aws-dashboard-console-using-arduino.md]: where -> were

**Pull request merged**